README.md
Preferenza
=======

Utility to store and consult preferences on Android. It allows to use several XML files for a better distribution of the information.

[![JitPack](https://jitpack.io/v/jdsdhp/preferenza.svg)](https://jitpack.io/#jdsdhp/preferenza) 
[![API](https://img.shields.io/badge/API-17%2B-red.svg?style=flat)](https://android-arsenal.com/api?level=17) 
[![License](https://img.shields.io/badge/License-MIT-blue.svg)](https://github.com/jdsdhp/preferenza/blob/master/LICENSE) 
[![Twitter](https://img.shields.io/badge/Twitter-@jdsdhp-9C27B0.svg)](https://twitter.com/jdsdhp)

## Including in your project

#### Gradle

```gradle
allprojects  {
    repositories {
        ...
        maven { url 'https://jitpack.io' }
    }
}
dependencies {
    implementation 'com.github.jdsdhp:preferenza:$version'
}
```

## Usage

### Kotlin
Everything will be saved and recovered as string.

```kotlin
storePreference(
    context = this,
    generalKey = "GENERAL_LIST_KEY",
    map = hashMapOf(
         "PARAM_1_KEY" to "Havana City", // "Havana City"
         "PARAM_2_KEY" to true,  // true.toString() = "true"
         "PARAM_3_KEY" to 666,   // 666.toString() = "666"
         "PARAM_4_KEY" to Any().toString()   // Any().toString() = "object identifier in this case"
    )
)

val havanaCity = consultPreference(
    this,
    "GENERAL_LIST_KEY",
    "PARAM_1_KEY",
    ""
)
Log.d("LOG", "CITY = $havanaCity") // Show "CITY = Havana City"

val numberRecovered = consultPreference(
    this,
    "GENERAL_LIST_KEY",
    "PARAM_3_KEY",
    ""
)
Log.d("LOG", "SUM = ${numberRecovered.toInt() + 2}") // Show "SUM = 668"
```

## Sample project

It's very important to check out the sample app. Most techniques that you would want to implement are already implemented in the examples.

View the sample app's source code [here](https://github.com/jdsdhp/preferenza/tree/master/app)

License
=======

    Copyright (c) 2020 jesusd0897.
    
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
    
        http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
Change Log
==========

Version 0.1.0 *(2020-07-17)*
----------------------------

Initial release.



preferenza/src/main/java/com/jesusd0897/preferenza/Preferences.kt
/*
 * Copyright (c) 2020 jesusd0897.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.jesusd0897.preferenza

import android.content.Context

/**
 * Save a list of preferences identified by a general key in a different xml file.
 * @param context    App context.
 * @param generalKey General key for a preferences list.
 * @param map    Map of preferences.
 */
fun storePreference(context: Context, generalKey: String, map: Map<String, Any?>) {
    val preferences = context.getSharedPreferences(generalKey, Context.MODE_PRIVATE)
    val edit = preferences.edit()
    if (map.isNotEmpty()) for ((key, value) in map) edit.putString(key, value.toString()).apply()
}

/**
 * Get a stored preference.
 * @param context    App context.
 * @param generalKey General key of preferences list.
 * @param keyValue   Specific key of preference.
 * @param defaultValue   Default value for return if nothing was found.
 * @return Value of a preference.
 */
fun consultPreference(
    context: Context,
    generalKey: String,
    keyValue: String,
    defaultValue: String
) =
    context.getSharedPreferences(generalKey, Context.MODE_PRIVATE)
        .getString(keyValue, defaultValue)!!

/**
 * Remove all custom saved preferences.
 * @param context App context.
 * @param generalKey General key of preferences.
 */
fun removePreference(context: Context, generalKey: String) =
    context.getSharedPreferences(generalKey, Context.MODE_PRIVATE).edit().clear().apply()

app/src/main/java/com/jesusd0897/preferenza/sample/MainActivity.kt
/*
 * Copyright (c) 2020 jesusd0897.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.jesusd0897.preferenza.sample

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import com.jesusd0897.preferenza.consultPreference
import com.jesusd0897.preferenza.storePreference

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //Everything will be saved as string
        storePreference(
            context = this,
            generalKey = "GENERAL_LIST_KEY",
            map = hashMapOf(
                "PARAM_1_KEY" to "Havana City", // "Havana City"
                "PARAM_2_KEY" to true,  // true.toString() = "true"
                "PARAM_3_KEY" to 666,   // 666.toString() = "666"
                "PARAM_4_KEY" to Any().toString()   // Any().toString() = "object identifier in this case"
            )
        )

        val havanaCity = consultPreference(
            this,
            "GENERAL_LIST_KEY",
            "PARAM_1_KEY",
            ""
        )
        Log.d("LOG", "CITY = $havanaCity") // Show "CITY = Havana City"

        val numberRecovered = consultPreference(
            this,
            "GENERAL_LIST_KEY",
            "PARAM_3_KEY",
            ""
        )
        Log.d("LOG", "SUM = ${numberRecovered.toInt() + 2}") // Show "SUM = 668"
    }
}